jobs: TestRegistryLifecycle failed

jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14572318?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14572318?buildTab=artifacts#/) on master @ [393aa0cd5acfa9fd8deab83e201b791f0bd61a98](https://github.com/cockroachdb/cockroach/commits/393aa0cd5acfa9fd8deab83e201b791f0bd61a98):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRegistryLifecycle (14m55s)
	TestRegistryLifecycle/dump_traces_on_pause-unpause-success (14m40s)

```
Stack: 

```
goroutine 34535 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* 
* goroutine 32346 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:423 +0x7c
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 32294
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:420 +0x2ec
* 
* goroutine 32362 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x6b16e28?, 0x400adac3f0?})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:452 +0x80
* runtime/pprof.Do({0x6b16a98?, 0xa3018a0?}, {{0x4000df7ac0?, 0x400169dce0?, 0x100d2a0?}}, 0x40031dd798)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x40031dd788?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:450 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 32294
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:442 +0xe4
* 
* goroutine 32311 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x6b16e28?, 0x400aa8f9e0?})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:452 +0x80
* runtime/pprof.Do({0x6b16a98?, 0xa3018a0?}, {{0x4000df7ac0?, 0x4001ea0f80?, 0x100d2a0?}}, 0x40031db798)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x40031db788?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:450 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 32294
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:442 +0xe4
* 
* goroutine 32342 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:423 +0x7c
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 32294
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:420 +0x2ec
* 
* goroutine 32343 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:500 +0x140
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 32294
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:471 +0x5c
* 
* goroutine 32302 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x6b16e28?, 0x400aa8f320?})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:452 +0x80
* runtime/pprof.Do({0x6b16a98?, 0xa3018a0?}, {{0x4000df7ac0?, 0x4001ea0c60?, 0x100d2a0?}}, 0x4003b65f98)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4003b65f88?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:450 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 32294
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:442 +0xe4
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/jobs @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37099

--- a/pkg/jobs/adopt.go
+++ b/pkg/jobs/adopt.go
@@ -89,7 +89,7 @@ func (r *Registry) maybeDumpTrace(resumerCtx context.Context, resumer Resumer, j
 	if err := r.db.Txn(dumpCtx, func(ctx context.Context, txn isql.Txn) error {
 		return WriteProtobinExecutionDetailFile(dumpCtx, resumerTraceFilename, &td, txn, jobID)
 	}); err != nil {
-		log.Warning(dumpCtx, "failed to write trace on resumer trace file")
+		log.Warningf(dumpCtx, "failed to write trace on resumer trace file: %v", err)
 	}
 }
 
@@ -468,7 +468,7 @@ func (r *Registry) runJob(
 	r.maybeClearLease(job, err)
 	r.maybeDumpTrace(ctx, resumer, job.ID())
 	if r.knobs.AfterJobStateMachine != nil {
-		r.knobs.AfterJobStateMachine()
+		r.knobs.AfterJobStateMachine(job.ID())
 	}
 	return err
 }

--- a/pkg/jobs/jobs_test.go
+++ b/pkg/jobs/jobs_test.go
@@ -176,7 +176,7 @@ type registryTestSuite struct {
 
 	// afterJobStateMachine is invoked in the AfterJobStateMachine testing knob if
 	// non-nil.
-	afterJobStateMachine func()
+	afterJobStateMachine func(jobspb.JobID)
 
 	// Instead of a ch for success, use a variable because it can retry since it
 	// is in a transaction.
@@ -200,9 +200,9 @@ func (rts *registryTestSuite) setUp(t *testing.T) {
 			}
 			return nil
 		}
-		knobs.AfterJobStateMachine = func() {
+		knobs.AfterJobStateMachine = func(id jobspb.JobID) {
 			if rts.afterJobStateMachine != nil {
-				rts.afterJobStateMachine()
+				rts.afterJobStateMachine(id)
 			}
 		}
 		args.Knobs.JobsTestingKnobs = knobs
@@ -891,22 +891,20 @@ func TestRegistryLifecycle(t *testing.T) {
 	t.Run("dump traces on pause-unpause-success", func(t *testing.T) {
 		ctx := context.Background()
 		completeCh := make(chan struct{})
-		var blockAfterJobStateMachine atomic.Bool
-		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func() {
-			if blockAfterJobStateMachine.Load() {
-				completeCh <- struct{}{}
+		var expectedJobID atomic.Int64
+		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func(id jobspb.JobID) {
+			if int64(id) != expectedJobID.Load() {
+				return
 			}
+			completeCh <- struct{}{}
 		}}
 		rts.setUp(t)
 		defer rts.tearDown()
-
-		blockAfterJobStateMachine.Store(true)
 		pauseUnpauseJob := func(expectedNumFiles int) {
-			j, err := jobs.TestingCreateAndStartJob(context.Background(), rts.registry, rts.idb(), rts.mockJob)
-			if err != nil {
-				t.Fatal(err)
-			}
+			j, err := jobs.TestingCreateAndStartJob(ctx, rts.registry, rts.idb(), rts.mockJob)
+			require.NoError(t, err)
 			rts.job = j
+			expectedJobID.Store(int64(j.ID()))
 
 			rts.mu.e.ResumeStart = true
 			rts.resumeCheckCh <- struct{}{}
@@ -938,22 +936,21 @@ func TestRegistryLifecycle(t *testing.T) {
 	t.Run("dump traces on fail", func(t *testing.T) {
 		ctx := context.Background()
 		completeCh := make(chan struct{})
-		var blockAfterJobStateMachine atomic.Bool
-		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func() {
-			if blockAfterJobStateMachine.Load() {
-				completeCh <- struct{}{}
+		var expectedJobID atomic.Int64
+		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func(id jobspb.JobID) {
+			if int64(id) != expectedJobID.Load() {
+				return
 			}
+			completeCh <- struct{}{}
 		}}
 		rts.setUp(t)
 		defer rts.tearDown()
 
-		blockAfterJobStateMachine.Store(true)
 		runJobAndFail := func(expectedNumFiles int) {
 			j, err := jobs.TestingCreateAndStartJob(ctx, rts.registry, rts.idb(), rts.mockJob)
-			if err != nil {
-				t.Fatal(err)
-			}
+			require.NoError(t, err)
 			rts.job = j
+			expectedJobID.Store(int64(j.ID()))
 
 			rts.mu.e.ResumeStart = true
 			rts.resumeCheckCh <- struct{}{}
@@ -978,22 +975,20 @@ func TestRegistryLifecycle(t *testing.T) {
 
 	t.Run("dump traces on cancel", func(t *testing.T) {
 		completeCh := make(chan struct{})
-		var blockAfterJobStateMachine atomic.Bool
-		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func() {
-			if blockAfterJobStateMachine.Load() {
-				completeCh <- struct{}{}
+		var expectedJobID atomic.Int64
+		rts := registryTestSuite{traceRealSpan: true, afterJobStateMachine: func(id jobspb.JobID) {
+			if int64(id) != expectedJobID.Load() {
+				return
 			}
+			completeCh <- struct{}{}
 		}}
 		rts.setUp(t)
 		defer rts.tearDown()
 
-		blockAfterJobStateMachine.Store(true)
 		j, err := jobs.TestingCreateAndStartJob(rts.ctx, rts.registry, rts.idb(), rts.mockJob)
-		if err != nil {
-			t.Fatal(err)
-		}
+		require.NoError(t, err)
 		rts.job = j
-
+		expectedJobID.Store(int64(j.ID()))
 		rts.mu.e.ResumeStart = true
 		rts.resumeCheckCh <- struct{}{}
 		rts.check(t, jobs.StatusRunning)

--- a/pkg/jobs/registry_test.go
+++ b/pkg/jobs/registry_test.go
@@ -724,7 +724,7 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 		jobMetrics               *JobTypeMetrics
 		adopted                  *metric.Counter
 		resumed                  *metric.Counter
-		afterJobStateMachineKnob func()
+		afterJobStateMachineKnob func(id jobspb.JobID)
 		// expectImmediateRetry is true if the test should expect immediate
 		// resumption on retry, such as after pausing and resuming job.
 		expectImmediateRetry bool
@@ -899,7 +899,7 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 	t.Run("running", func(t *testing.T) {
 		ctx := context.Background()
 		bti := BackoffTestInfra{}
-		bti.afterJobStateMachineKnob = func() {
+		bti.afterJobStateMachineKnob = func(_ jobspb.JobID) {
 			if bti.done.Load().(bool) {
 				return
 			}
@@ -928,7 +928,7 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 			bti.errCh <- nil
 			<-bti.transitionCh
 		}
-		bti.afterJobStateMachineKnob = func() {
+		bti.afterJobStateMachineKnob = func(jobspb.JobID) {
 			if bti.done.Load().(bool) {
 				return
 			}
@@ -954,7 +954,7 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 	t.Run("revert on fail", func(t *testing.T) {
 		ctx := context.Background()
 		bti := BackoffTestInfra{}
-		bti.afterJobStateMachineKnob = func() {
+		bti.afterJobStateMachineKnob = func(jobspb.JobID) {
 			if bti.done.Load().(bool) {
 				return
 			}
@@ -1012,7 +1012,7 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 			bti.errCh <- nil
 			<-bti.transitionCh
 		}
-		bti.afterJobStateMachineKnob = func() {
+		bti.afterJobStateMachineKnob = func(jobspb.JobID) {
 			if bti.done.Load().(bool) {
 				return
 			}

--- a/pkg/jobs/testing_knobs.go
+++ b/pkg/jobs/testing_knobs.go
@@ -65,7 +65,7 @@ type TestingKnobs struct {
 	// AfterJobStateMachine is called once the running instance of the job has
 	// returned from the state machine that transitions it from one state to
 	// another.
-	AfterJobStateMachine func()
+	AfterJobStateMachine func(jobspb.JobID)
 
 	// TimeSource replaces registry's clock.
 	TimeSource *hlc.Clock

